src/trigger/HSISourceModel.hpp

```diff
@@ -0,0 +1,122 @@
+#include "appdal/HSI2TCTranslatorConf.hpp" 
+#include "appdal/HSISignalWindow.hpp" 
+
+namespace dunedaq::trigger {
+
+
+class HSISourceModel : public readoutlibs::SourceConcept
+{
+public: 
+  using inherited = readoutlibs::SourceConcept;
+
+  /**
+   * @brief SourceModel Constructor
+   * @param name Instance name for this SourceModel instance
+   */
+  
+  HSISourceModel(): readoutlibs::SourceConcept() {}
+  ~HSISourceModel() {}
+
+  void init(const coredal::DaqModule* cfg) override {
+    if (cfg->get_outputs().size() != 1) {
+      throw readoutlibs::InitializationError(ERS_HERE, "Only 1 output supported for subscribers");
+    }
+    m_data_sender = get_iom_sender<trigger::TCWrapper>(cfg->get_outputs()[0]->UID());
+
+    if (cfg->get_inputs().size() != 1) {
+      throw readoutlibs::InitializationError(ERS_HERE, "Only 1 input supported for subscribers");
+    }
+    m_data_receiver = get_iom_receiver<dfmessages::HSIEvent>(cfg->get_inputs()[0]->UID());
+    auto data_reader = cfg->cast<appdal::DataSubscriber>();
+    if (data_reader == nullptr) {
+       throw readoutlibs::InitializationError(ERS_HERE, "DAQ module is not a DataReader");
+    }
+    auto hsi_conf = data_reader->get_configuration()->cast<appdal::HSI2TCTranslatorConf>();
+    if (hsi_conf == nullptr) {
+	throw readoutlibs::InitializationError(ERS_HERE, "Missing HSI2TCTranslatorConf");
+    }
+    for (auto win : hsi_conf->get_signals()) {
+	   m_signals[win->get_signal_type()] = std::pair<uint32_t, uint32_t>(win->get_time_before(), win->get_time_after());
+    }
+  }
+
+  void start() {
+    m_data_receiver->add_callback(std::bind(&HSISourceModel::handle_payload, this, std::placeholders::_1));
+  }  
+
+  void stop() {
+    m_data_receiver->remove_callback();
+  }
+
+  void get_info(opmonlib::InfoCollector& /*ci*/, int /*level*/) {}
+
+  bool handle_payload(dfmessages::HSIEvent& data) // NOLINT(build/unsigned)
+  {
+    
+    triggeralgs::TriggerCandidate candidate;
+    auto signal_info = m_signals.find(data.signal_map);
+    if (signal_info != m_signals.end()) {
+        // clang-format off
+        candidate.time_start = data.timestamp - signal_info->second.first;  // time_start
+        candidate.time_end   = data.timestamp + signal_info->second.second; // time_end,
+        // clang-format on    
+    } else {
+        throw dunedaq::trigger::SignalTypeError(ERS_HERE, "HSI subscriber" , data.signal_map);
+    }
+    
+    candidate.time_candidate = data.timestamp;
+    // throw away bits 31-16 of header, that's OK for now
+    candidate.detid = (uint)detdataformats::DetID::Subdetector::kDAQ ; // NOLINT(build/unsigned)
+    candidate.type = triggeralgs::TriggerCandidate::Type::kTiming;
+
+    candidate.algorithm = triggeralgs::TriggerCandidate::Algorithm::kHSIEventToTriggerCandidate;
+    candidate.inputs = {};
+
+    trigger::TCWrapper tc(candidate);
+    if (!m_data_sender->try_send(std::move(tc), iomanager::Sender::s_no_block)) {
+      ++m_dropped_packets;
+    }
```

Comment:
I think the normal logic for our HSI handler is that it keeps on trying to send the TC (in a while loop) until it is sent, so no TCs are dropped at this stage (but they can get "stuck" in a while loop if we don't have the output queue).

Reply:
and if they get very old this all becomes a mess... :-)